### ISSUE TITLE
The behavior of "Run to user code"

### ISSUE THREAD

**Issue by wk-952** [x64dbg/x64dbg#1489]
Almost every time whenever the option `Run to user code` is used and `EIP` is inside a windows function, the debugger lands at a code with a long-jump instruction (similar to WOW64 transition) rather than at the actual program code:
![1](https://cloud.githubusercontent.com/assets/11916365/23807752/8df8c7da-05cf-11e7-9aae-181b6c610d5b.png)
eventually you'll have to press `Alt + F9` (`Run to user code`) a couple of times to reach the actual user code.
Is there any way to change/fix this ?

- Steps to reproduce:
1. Debug the provided application (simple hello world application)
2. Step-in 3 times until you're at the very first instruction of `printf`
3. Press `Alt + F9` (`Run to user code`) and you'll land at the long-jmp instruction
4. To reach the actual user code, press `Alt + F9` ~4 times (2nd time is a little bit slow)

- This issue is reproducible on 32-bit applications (i don't remember facing this issue on any 64-bit application but of course i may be wrong)
- Reproduced on Win10 x64
- Change extension to `.exe` :smile: 
[hello_world_tst3.txt](https://github.com/x64dbg/x64dbg/files/834875/hello_world_tst3.txt)

**Comment by wk-952** [comment]
Here's a 64-bit version of the application, try the same steps and this issue won't be triggered
[hello_world_tst3_x64.txt](https://github.com/x64dbg/x64dbg/files/834919/hello_world_tst3_x64.txt)

**Comment by mrexodia** [comment]
Hey, yeah I saw the issue before. The reason this happens is that `Run to user code` will actually execute the command `TraceIntoConditional !mod.party(cip)`. Since the wow64 code is not associated with a module it will kinda fail there. I guess there are two possible solutions:

1. Change the condition to explicitly mention the `ljmp` and ignore that
2. Change to `TraceOverConditional`
3. Add the wow64 pages as modules

Downsides:

1. Will work with standard code but would fail if the syscalls are hooked (ScyllaHide)
2. Won't work in certain cases (eg abnormal control flow in system structure)
3. Undocumented so might be annoying.

**Comment by wk-952** [comment]
Number 2 seemed fair enough for me, also as you said this wow64 code isn't associated with a module, so based on that i did try the following as a stop condition for the `TraceIntoConditional`:
`!mod.party(cip) && mod.base(cip) && mod.entry(cip)`
since both `mod.base(cip)` and `mod.entry(cip)` always returns 0 in this case.
It worked in some cases, but in others it didn't; sometimes even the user code would be completely bypassed.

To demonstrate this,
1. Download this [hello_world](https://github.com/x64dbg/x64dbg/files/836031/hello_world_tst3_msgbox.txt) app (same one as above but with a MessageBox) and pause the debugger when the message box is displayed
2. In the command bar type:
`TraceIntoConditional !mod.party(cip) && mod.base(cip) && mod.entry(cip)`
then press ok in the message box
3. The dubugee will be terminated even though there are still some code after the call to `MessageBoxA()`
4. Again debug this app but this time use `TraceOverConditional` instead of `TraceIntoConditional`
and the debugger will pause at the proper code as expected.

Am i doing something wrong ?

**Comment by mrexodia** [comment]
The reason was the WOW64 layer. If the instruction executed is `ljmp 33,XXXXXXXX` to switch to 64 bit the stepping will be messed up for some reason... I made the requirements for the workaround less strict and `TraceIntoConditional` now works fine (it will terminate after 50k steps on clicking OK). I switched to `TraceOverConditional` because it's better in pretty much every way.

Thanks a lot,

Duncan
